### PR TITLE
Documentation Refactor Part 1

### PR DESCRIPTION
**Pull request by anthony-santana**
### Description

What this PR does do:
* Refactors the entire docs index structure
    * "Tutorials" -> "Applications"
    * New sections that contain existing docs:
        * Backends
        * Installation
        * Integration
        * Extending
* Creates a new Quick Start Guide and Basics section
* Refactors the cuda-quantum/docs folder to be more similar to index structure
* Reproduces examples in C++ and Python, and moves them to alternate tabs

What this PR doesn't do:
* Add any documentation for new python support
* Make any large modifications to examples/applications content
* Make any large modifications to existing content in the bottom half of the index structure

### CODE REVIEW

File: docs/sphinx/quick_start.rst
Comment (mmvandieren) [inline docs/sphinx/quick_start.rst:1]: line 9: system-level should be hyphenated

Comment (mmvandieren) [inline docs/sphinx/quick_start.rst:1]: to break up the long 2nd sentence, I recommend changing the comma in line 8 after 'together' with a period.  Then the rest of the paragraph could be:
Further enabling application acceleration, CUDA Quantum includes language extensions for Python and C++ and a system-level toolchain.

Comment (mmvandieren) [inline docs/sphinx/quick_start.rst:1]: the emdash on line 8 doesn't seem to render as such on the docs page.  I think we need to use `&mdash;`

File: docs/sphinx/using/install/data_center_install.rst [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Installation from Source
-*******************************************
++++++++++++++++++++++++++
```
Comment (bettinaheim) [inline docs/sphinx/using/install/data_center_install.rst:2]: I think this will change the hierarchy of the page, such that subsequent sections are no longer subsections under this title. Was that intentional?

File: docs/sphinx/using/install/install.rst [hunk 2]
```diff
@@ -893,20 +893,27 @@ you can set the size of the secret string to up to 28-32 when running on the `nv
...
 When you execute this command, the program simply seems to hang; that is because it takes
 a long time for the CPU-only backend to simulate 28+ qubits! Cancel the execution with `Ctrl+C`.
 
 You are now all set to start developing quantum applications using CUDA Quantum!
-Please proceed to :doc:`Using CUDA Quantum <using/cudaq>` to learn the basics.
+Please proceed to :doc:`Basics <../basics>` for an introduction
+to the fundamental features of CUDA Quantum.
+
+
+Data Center Installation
```
Comment (bettinaheim) [inline docs/sphinx/using/install/install.rst:916]: Ah, that explains why you changes the section hierarchy in the other document. I think the Installation from source needs to be a separate page, since the vast majority of the people should never look at it, but a good number of people sooner or later should read the install page. I don't think this is a good page to absorb this into one gigantic page.

Comment (anthony-santana) [reply docs/sphinx/using/install/install.rst]: I've updated the index structure as follows:

```
> Quick Start
.
.
.
> Installation
      >> Local Installation
            >>>> Introduction
            >>>> ...
      >> Data Center Installation
            >>>> Installation from Source
            >>>> ...
```

File: docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// If you have a NVIDIA GPU you can use this example to see
+// that the GPU-accelerated backends can easily handle a
+// larger number of qubits compared the CPU-only backend.
+
+// Depending on the available memory on your GPU, you can
+// set the number of qubits to around 30 qubits, and set the
+// `--target=nvidia` from the command line.
+
+// Note: Without setting the target to the `nvidia` backend,
+// a 30 qubit simulation simply seems to hang; that is
```
Comment (khalatepradnya) [inline docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp:13]: In PR#1295, there was a discussion about the term `hang`...
Ref: https://github.com/NVIDIA/cuda-quantum/pull/1295#discussion_r1504518984

File: docs/sphinx/quick_start.rst [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+If you want to develop CUDA Quantum applications using Python, install the
+latest stable release of the CUDA Quantum Python API:  
+
+.. code-block:: bash
+
+    pip install cuda-quantum
+
+For further information, see the `CUDA Quantum project <https://pypi.org/project/cuda-quantum/>`_ on PyPI.
+
+We also offer a fully featured CUDA Quantum installation, including all C++ and Python tools, via our
```
Comment (khalatepradnya) [inline docs/sphinx/quick_start.rst:31]: Consistency in listing languages - I see that now `Python` is called out to be first elsewhere, so perhaps mention `Python and C++` everywhere?

File: docs/sphinx/index.rst [hunk 5]
```diff
@@ -11,14 +11,14 @@ You are browsing the documentation for |version| version of CUDA Quantum. You ca
...
+      Examples <using/examples/examples.rst>
+      Applications <using/tutorials.rst>
+      Backends <using/backends/backends.rst>
+      Installation <using/install/install.rst>
+      Integration <using/integration/integration.rst>
+      Extending <using/extending/extending.rst>
       Specifications <specification/index.rst>
       API Reference <api/api.rst>
-      Data Center Installation Guide <data_center_install.rst>
       Other Versions <versions.rst>
```
Comment (khalatepradnya) [inline docs/sphinx/index.rst:24]: I see that the `Other Versions` link was not changed, however, using the latest docs from CI -https://htmlpreview.github.io/?https%3A%2F%2Fgithub.com%2FNVIDIA%2Fcuda-quantum%2Fblob%2Fdocs%2Fpreview%2Fpr-1225%2Findex.html= , the other versions link isn't rendering, it shows this - https://[raw.githubusercontent.com/NVIDIA/cuda-quantum/docs/preview/latest/releases.html](https://raw.githubusercontent.com/NVIDIA/cuda-quantum/docs/preview/latest/releases.html)

File: docs/sphinx/quick_start.rst [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+in quantum algorithm research. It offers a unified programming model designed for a hybrid
+setting |---| that is, CPUs, GPUs, and QPUs working together. Further enabling application
+acceleration, CUDA Quantum includes language extensions for Python and C++ and a system-level toolchain.
+
+.. 
+    Fix Me: Learn more about CUDA Quantum’s key benefits here [Link to CUDA Quantum Marketing page]
+
+This Quick Start page guides you through getting set up with CUDA Quantum and running your first program.
+If you have already installed and configured CUDA Quantum, we encourage you to move directly to our
+:ref:`Basics Section <cudaq-basics-landing-page>`.
```
[the remark this answers is not in the document]
Comment (anthony-santana) [reply docs/sphinx/quick_start.rst]: I will update these but I think the ref is advantageous in that you don't need to hard-code paths to these documents. If anyone moves any of these files around in the future, they will have to go in again and fix all of the paths manually.

[the remark this answers is not in the document]
Comment (anthony-santana) [reply docs/sphinx/quick_start.rst]: I am currently that someone in the future and this is not fun to work with... As I'm making changes per the other PR comments, thus moving files around, these links repeatedly need updating. I am heavily in favor of using ref's.

File: docs/sphinx/using/examples/hardware_providers.rst [hunk 7]
```diff
@@ -0,0 +1,37 @@
+.. _python-examples-for-hardware-providers:
+.. _cpp-examples-for-hardware-providers:
+
+Using Quantum Hardware Providers
+-----------------------------------
+
+CUDA Quantum contains support for using a set of hardware providers (Quantinuum, IonQ, and IQM).
```
[the remark this answers is not in the document]
Comment (anthony-santana) [reply docs/sphinx/using/examples/hardware_providers.rst:7]: Not sure as I didn't write any of this content originally 🤷  Fixing it now

File: docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+  // Apply another set of Hadamards to the qubits.
+  h(qvector);
+
+  // Apply measurement gates to just the `qubits`
+  // (excludes the auxillary qubit).
+  mz(qvector);
+}
+
+int main() {
```
Comment (bettinaheim) [inline docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp:62]: This example is referenced in the installation guide, and since you remove the command line handling, th e commands in the install guide now won't work. Could you please undo the changes to this example?

File: docs/sphinx/index.rst [hunk 9]
```diff
@@ -3,22 +3,31 @@ CUDA Quantum
 ************
 
 Welcome to the CUDA Quantum documentation page!
-You are browsing the documentation for |version| version of CUDA Quantum. You can find documentation for all released versions :doc:`here <versions>`.
+
+CUDA Quantum streamlines hybrid application development and promotes productivity and scalability
+in quantum algorithm research. It offers a unified programming model designed for a hybrid
```
Comment (bettinaheim) [inline docs/sphinx/index.rst:8]:
```suggestion
in quantum computing. It offers a unified programming model designed for a hybrid
```

File: docs/sphinx/using/backends/backends.rst [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+   :caption: Backend Targets
+   :maxdepth: 1
+
+      Simulation <simulators.rst>
+      Hardware <hardware.rst>
+
+**The following is a comprehensive list of the available targets in CUDA Quantum:**
+
+.. 
+   Fix Me: Add links to all of the following
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply docs/sphinx/using/backends/backends.rst:15]: Why was this resolved? I don't see any links.

File: docs/sphinx/using/basics/basics.rst [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+      What is a CUDA Quantum Kernel? <kernel_intro.rst>
+      Building your first CUDA Quantum Program <build_kernel.rst>
+      Running your first CUDA Quantum Program <run_kernel.rst>
+      Troubleshooting <troubleshooting.rst>
+
+..
+  Fix Me: Move to own file when ready to fill out section (Issue #1352)
+
+  Language Fundamentals
```
Comment (bettinaheim) [inline docs/sphinx/using/basics/basics.rst:16]: Please remove this section completely. It is not written, so it should not be here. The documentation will be live update when we merge this change to main; there certainly shouldn't be a Todo visible in the docs, or any empty/half finished paragraphs.

File: docs/sphinx/using/examples/hardware_providers.rst [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+   .. literalinclude:: ../../examples/python/providers/ionq.py
+      :language: python
+
+.. tab:: C++
+
+   .. literalinclude:: ../../examples/cpp/providers/ionq.cpp
+      :language: cpp
+
+The following code illustrates how to run kernels on IQM's backends.
```
Comment (bettinaheim) [inline docs/sphinx/using/examples/hardware_providers.rst:32]: Why is the OCQ example missing here?

File: docs/sphinx/using/examples/introduction.rst [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+   We're going to take a look at how to construct quantum programs through CUDA Quantum's `Kernel` API.
+
+   When you create a `Kernel` and invoke its methods, a quantum program is constructed that can then be executed by calling, for example, `cudaq::sample`. Let's take a closer look!
+
+   .. literalinclude:: ../../examples/python/intro.py
+      :language: python
+
+.. tab:: C++
+
+   We're going to take a look at how to construct quantum programs using CUDA Quantum kernel expressions.
```
Comment (bettinaheim) [inline docs/sphinx/using/examples/introduction.rst:17]:
```suggestion
   We're going to take a look at how to construct quantum programs using CUDA Quantum kernels.
```
quantum kernels are functions, and hence calling them expressions seems odd from a language perspective.

File: docs/sphinx/versions.rst [hunk 14]
```diff
@@ -1,6 +0,0 @@
-************************
```
Comment (bettinaheim) [inline docs/sphinx/versions.rst:1]: Why was this removed? This file is needed for the docs versioning to work correctly.

File: docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp (hunk 8 above)
Comment (anthony-santana) [reply docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp:62]: I didn't change the existing Bernstein Vazirani file in examples/algorithms , the one which is referenced in the installation guide. This is a new BV algorithm written specifically for the examples, designed to match side-by-side with the python version.

Comment (bettinaheim) [reply docs/sphinx/examples/cpp/basics/bernstein_vazirani.cpp:62]: ah, ok. Since we include the examples folder in the docker image, we should move one of them out of that folder and e.g. into the snippets. I'll do that with the updates to the install guide.

File: docs/sphinx/using/backends/backends.rst [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+      Simulation <simulators.rst>
+      Hardware <hardware.rst>
+      Multi-Processor Platforms <platform.rst>
+
+**The following is a comprehensive list of the available targets in CUDA Quantum:**
+
+* :ref:`density-matrix-cpu <default-simulator>`
+* :ref:`ionq <ionq-backend>`
+* :ref:`iqm <iqm-backend>`
+* :code:`nvcf`
```
Comment (bettinaheim) [inline docs/sphinx/using/backends/backends.rst:17]:
```suggestion
```

File: docs/sphinx/using/backends/backends.rst [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+* :ref:`nvidia <nvidia-backend>`
+* :ref:`nvidia-fp64 <nvidia-fp64-backend>`
+* :ref:`nvidia-mqpu <nvidia-mgpu-backend>`
+* :ref:`nvidia-mqpu-fp64 <nvidia-mgpu-backend>`
+* :ref:`oqc <oqc-backend>`
+* :code:`orca`
```
Comment (bettinaheim) [inline docs/sphinx/using/backends/backends.rst:23]:
```suggestion
```
@anthony-santana Please create a github issue to track the missing docs for orca and photonics

File: docs/sphinx/using/backends/backends.rst [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+* :code:`photonics`
```
Comment (bettinaheim) [inline docs/sphinx/using/backends/backends.rst:24]:
```suggestion
```

File: docs/sphinx/using/backends/backends.rst [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+* :ref:`qpp-cpu <qpp-cpu-backend>`
+* :ref:`quantinuum <quantinuum-backend>`
+* :code:`remote-mqpu`
```
Comment (bettinaheim) [inline docs/sphinx/using/backends/backends.rst:27]: This is documented in docs/sphinx/using/backends/platform.rst

**Review by bettinaheim — APPROVED** [review]
Added some comments to update or remove the backend links, and redirected this into the feature branch. Other than that, good to go into the feature branch.

File: docs/sphinx/using/basics/kernel_intro.rst [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+standard quantum circuit. We can not only parameterize the kernel, but can also contain classical control
+flow statements (`if`, `for`, `while`, etc.), and classical computations such as additions, multiplication, etc.
+Conditional statements on quantum memory and qubit measurements can be included in quantum kernels to enable 
+dynamic circuits and fast feedback, particularly useful for quantum error correction. To learn more about what
+language constructs are supported within quantum kernels, take a look at the CUDA Quantum 
+:doc:`specification <../../specification/cudaq/kernels>`.
+
+**How do I build and run a quantum kernel?**
+
+Once a quantum kernel has been defined in a program, it can be executed using the `sample` or the `observe` primitives.
```
Comment (amccaskey) [inline docs/sphinx/using/basics/kernel_intro.rst:25]: It can also just be called.

File: docs/sphinx/using/basics/run_kernel.rst [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+
+  .. literalinclude:: ../../snippets/python/using/time.py
+        :language: python
+        :start-after: [Begin Time]
+        :end-before: [End Time]
+
+
+.. tab:: C++
+
+  Using the `-- target` argument to `nvq++`, different targets can be specified for kernel execution.
```
Comment (amccaskey) [inline docs/sphinx/using/basics/run_kernel.rst:191]: extra space in --target?

File: docs/sphinx/using/backends/backends.rst (hunk 16 above)
Comment (anthony-santana) [reply docs/sphinx/using/backends/backends.rst:23]: Issue #1388 and #1389 have been filed and I've removed them from this list
